coderd/coderdtest/oidctest/idp.go

```diff
@@ -58,15 +59,107 @@ type deviceFlow struct {
 	granted   bool
 }
 
+type fakeIDPProtected struct {
+	mu sync.RWMutex
+
+	issuer     string
+	issuerURL  *url.URL
+	key        *rsa.PrivateKey
+	provider   ProviderJSON
+	handler    http.Handler
+	cfg        *oauth2.Config
+	fakeCoderd func(req *http.Request) (*http.Response, error)
+}
+
+func (f *fakeIDPProtected) Issuer() string {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+	return f.issuer
+}
+
+func (f *fakeIDPProtected) IssuerURL() *url.URL {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+	return f.issuerURL
+}
+
+func (f *fakeIDPProtected) PrivateKey() *rsa.PrivateKey {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+	return f.key
+}
+
+func (f *fakeIDPProtected) Provider() ProviderJSON {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+	return f.provider
+}
+
+func (f *fakeIDPProtected) Config() *oauth2.Config {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+	return f.cfg
+}
+
+func (f *fakeIDPProtected) Handler() http.Handler {
+	f.mu.RLock()
+	defer f.mu.RUnlock()
+	return f.handler
+}
+
+func (f *fakeIDPProtected) SetIssuer(issuer string) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	f.issuer = issuer
+}
+
+func (f *fakeIDPProtected) SetIssuerURL(issuerURL *url.URL) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	f.issuerURL = issuerURL
+}
+
+func (f *fakeIDPProtected) SetProvider(provider ProviderJSON) {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	f.provider = provider
+}
+
+func (f *fakeIDPProtected) MutateConfig(fn func(cfg *oauth2.Config)) {
+	f.mu.RLock()
```

Comment:
Nit: Would be simpler to just use Lock immediately here.

Reply:
The only problem here is that if someone tries to do a re-entrant lock the test will fail. What I can do instead is modify the locked fields to use `TryLock()` and panic if it fails. I think this is actually legit for something that is only to be used in tests.

Reply:
Yeah you're right. I reverted that change.